File: sdk-typescript/src/governance/strict-receipt-prepared-state.ts
export const DEFINITIVE_NO_STORE = Object.freeze({
  status: 'definitive_no_store' as const,
});

export type DefinitiveNoStore = typeof DEFINITIVE_NO_STORE;
export type PreparedReceiptKind = 'decision' | 'resolution' | 'timeout';

export interface PreparedReceiptView<T> {
  token: string;
  receipt_hash: string;
  kind: PreparedReceiptKind;
  value: T;
}

interface PreparedReceipt<T> extends PreparedReceiptView<T> {
  fingerprint: string;
  commit: () => void;
}

export interface PreparedStateInspection {
  frozen: boolean;
  freeze_reason?: string;
  prepared?: Omit<PreparedReceiptView<unknown>, 'value'>;
}

export type PreparedReconciliation =
  | { status: 'stored'; token: string; receipt_hash: string }
  | { status: 'definitive_no_store'; token: string; receipt_hash: string;
    capability: DefinitiveNoStore }
  | { status: 'ambiguous'; token: string; receipt_hash: string; reason: string };

export class PreparedReceiptState {
  private prepared: PreparedReceipt<unknown> | undefined;
  private frozenReason: string | undefined;

  constructor(private readonly tokenFactory: () => string) {}

  retry<T>(
    fingerprint: string, kind: PreparedReceiptKind,
  ): PreparedReceiptView<T> | undefined {
    this.assertNotFrozen();
    if (!this.prepared) return undefined;
    if (this.prepared.fingerprint !== fingerprint || this.prepared.kind !== kind) {
      throw new Error('a different receipt is already prepared for this session');
    }
    return this.view(this.prepared as PreparedReceipt<T>);
  }

  prepare<T>(params: {
    fingerprint: string; receipt_hash: string; kind: PreparedReceiptKind;
    value: T; commit: () => void;
  }): PreparedReceiptView<T> {
    this.assertNotFrozen();
    if (this.prepared) throw new Error('a receipt is already prepared for this session');
    const token = this.tokenFactory();
    if (typeof token !== 'string' || token.trim().length === 0) {
      throw new Error('prepared token factory must return a nonblank string');
    }
    this.prepared = { token, ...params };
    return this.view(this.prepared as PreparedReceipt<T>);
  }

  commit<T>(token: string, receiptHash: string): T {
    const prepared = this.match(token, receiptHash);
    try {
      prepared.commit();
    } catch (error) {
      this.frozenReason = 'accepted_but_local_commit_failed';
      throw error;
    }
    this.prepared = undefined;
    this.frozenReason = undefined;
    return prepared.value as T;
  }

  abort(token: string, receiptHash: string, capability: DefinitiveNoStore): void {
    if (capability !== DEFINITIVE_NO_STORE) {
      throw new Error('abort requires the definitive_no_store capability');
    }
    this.match(token, receiptHash);
    this.prepared = undefined;
    this.frozenReason = undefined;
  }

  freeze(token: string, receiptHash: string, reason: string): void {
    this.match(token, receiptHash);
    if (typeof reason !== 'string' || reason.trim().length === 0) {
      throw new Error('freeze reason must be nonblank');
    }
    this.frozenReason = reason;
  }

  reconcile<T>(input: PreparedReconciliation): T | undefined {
    if (input.status === 'stored') return this.commit<T>(input.token, input.receipt_hash);
    if (input.status === 'definitive_no_store') {
      this.abort(input.token, input.receipt_hash, input.capability);
      return undefined;
    }
    this.freeze(input.token, input.receipt_hash, input.reason);
    return undefined;
  }

  inspect(): PreparedStateInspection {
    return {
      frozen: this.frozenReason !== undefined,
      ...(this.frozenReason === undefined ? {} : { freeze_reason: this.frozenReason }),
      ...(this.prepared === undefined ? {} : {
        prepared: {
          token: this.prepared.token, receipt_hash: this.prepared.receipt_hash,
          kind: this.prepared.kind,
        },
      }),
    };
  }

  reset(): void {
    this.prepared = undefined;
    this.frozenReason = undefined;
  }

  private assertNotFrozen(): void {
    if (this.frozenReason !== undefined) {
      throw new Error('strict receipt session is frozen pending reconciliation');
    }
  }

  private match(token: string, receiptHash: string): PreparedReceipt<unknown> {
    if (!this.prepared) throw new Error('no receipt is prepared');
    if (token !== this.prepared.token) throw new Error('prepared token mismatch');
    if (receiptHash !== this.prepared.receipt_hash) {
      throw new Error('prepared receipt hash mismatch');
    }
    return this.prepared;
  }

  private view<T>(prepared: PreparedReceipt<T>): PreparedReceiptView<T> {
    return {
      token: prepared.token, receipt_hash: prepared.receipt_hash,
      kind: prepared.kind, value: prepared.value,
    };
  }
}

File: sdk-python/obsvr/strict_receipt_prepared_state.py
"""Opaque two-phase state for process-local strict receipt coordination."""

from __future__ import annotations

from typing import Any, Callable, Dict, Optional


class _DefinitiveNoStore:
    status = "definitive_no_store"


DEFINITIVE_NO_STORE = _DefinitiveNoStore()


class PreparedReceiptState:
    """Hold at most one signed receipt until transport certainty exists."""

    def __init__(self, token_factory: Callable[[], str]) -> None:
        self._token_factory = token_factory
        self._prepared: Optional[Dict[str, Any]] = None
        self._frozen_reason: Optional[str] = None

    def retry(self, fingerprint: str, kind: str) -> Optional[Dict[str, Any]]:
        self._assert_not_frozen()
        if self._prepared is None:
            return None
        if (
            self._prepared["fingerprint"] != fingerprint
            or self._prepared["kind"] != kind
        ):
            raise ValueError(
                "a different receipt is already prepared for this session"
            )
        return self._view(self._prepared)

    def prepare(
        self, *, fingerprint: str, receipt_hash: str, kind: str,
        value: Any, commit: Callable[[], None],
    ) -> Dict[str, Any]:
        self._assert_not_frozen()
        if self._prepared is not None:
            raise ValueError("a receipt is already prepared for this session")
        token = self._token_factory()
        if not isinstance(token, str) or not token.strip():
            raise ValueError("prepared token factory must return a nonblank string")
        self._prepared = {
            "token": token, "receipt_hash": receipt_hash, "kind": kind,
            "value": value, "fingerprint": fingerprint, "commit": commit,
        }
        return self._view(self._prepared)

    def commit(self, token: str, receipt_hash: str) -> Any:
        prepared = self._match(token, receipt_hash)
        try:
            prepared["commit"]()
        except Exception:
            self._frozen_reason = "accepted_but_local_commit_failed"
            raise
        self._prepared = None
        self._frozen_reason = None
        return prepared["value"]

    def abort(self, token: str, receipt_hash: str, capability: Any) -> None:
        if capability is not DEFINITIVE_NO_STORE:
            raise ValueError("abort requires the definitive_no_store capability")
        self._match(token, receipt_hash)
        self._prepared = None
        self._frozen_reason = None

    def freeze(self, token: str, receipt_hash: str, reason: str) -> None:
        self._match(token, receipt_hash)
        if not isinstance(reason, str) or not reason.strip():
            raise ValueError("freeze reason must be nonblank")
        self._frozen_reason = reason

    def reconcile(self, input_value: Dict[str, Any]) -> Any:
        status = input_value.get("status")
        if status == "stored":
            return self.commit(input_value.get("token"), input_value.get("receipt_hash"))
        if status == "definitive_no_store":
            self.abort(
                input_value.get("token"), input_value.get("receipt_hash"),
                input_value.get("capability"),
            )
            return None
        if status == "ambiguous":
            self.freeze(
                input_value.get("token"), input_value.get("receipt_hash"),
                input_value.get("reason"),
            )
            return None
        raise ValueError("unsupported prepared reconciliation status")

    def inspect(self) -> Dict[str, Any]:
        result: Dict[str, Any] = {"frozen": self._frozen_reason is not None}
        if self._frozen_reason is not None:
            result["freeze_reason"] = self._frozen_reason
        if self._prepared is not None:
            result["prepared"] = {
                key: self._prepared[key]
                for key in ("token", "receipt_hash", "kind")
            }
        return result

    def reset(self) -> None:
        self._prepared = None
        self._frozen_reason = None

    def _assert_not_frozen(self) -> None:
        if self._frozen_reason is not None:
            raise ValueError("strict receipt session is frozen pending reconciliation")

    def _match(self, token: str, receipt_hash: str) -> Dict[str, Any]:
        if self._prepared is None:
            raise ValueError("no receipt is prepared")
        if token != self._prepared["token"]:
            raise ValueError("prepared token mismatch")
        if receipt_hash != self._prepared["receipt_hash"]:
            raise ValueError("prepared receipt hash mismatch")
        return self._prepared

    @staticmethod
    def _view(prepared: Dict[str, Any]) -> Dict[str, Any]:
        return {
            "token": prepared["token"], "receipt_hash": prepared["receipt_hash"],
            "kind": prepared["kind"], "value": prepared["value"],
        }
